istioctl/cmd/add-to-mesh.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+	"go.uber.org/multierr"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/istio/istioctl/pkg/util/handlers"
+	istiocmd "istio.io/istio/pilot/cmd"
+	"istio.io/istio/pilot/pkg/kube/inject"
+	"istio.io/pkg/log"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+)
+
+var (
+	// Function that creates Kubernetes client-go; making it a variable lets us mock client-go
+	interfaceFactory = createInterface
+)
+
+func addToMeshCmd() *cobra.Command {
+	addToMeshCmd := &cobra.Command{
+		Use:     "add-to-mesh",
+		Aliases: []string{"add"},
+		Short:   "Add workloads into Istio service mesh",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.HelpFunc()(cmd, args)
+			if len(args) != 0 {
+				return fmt.Errorf("unknown resource type %q", args[0])
+			}
+			return nil
+		},
+	}
+	addToMeshCmd.AddCommand(svcMeshifyCmd())
+	addToMeshCmd.PersistentFlags().StringVar(&meshConfigFile, "meshConfigFile", "",
+		"mesh configuration filename. Takes precedence over --meshConfigMapName if set")
+	addToMeshCmd.PersistentFlags().StringVar(&injectConfigFile, "injectConfigFile", "",
+		"injection configuration filename. Cannot be used with --injectConfigMapName")
+	addToMeshCmd.PersistentFlags().StringVar(&valuesFile, "valuesFile", "",
+		"injection values configuration filename.")
+
+	addToMeshCmd.PersistentFlags().StringVar(&meshConfigMapName, "meshConfigMapName", defaultMeshConfigMapName,
+		fmt.Sprintf("ConfigMap name for Istio mesh configuration, key should be %q", configMapKey))
+	addToMeshCmd.PersistentFlags().StringVar(&injectConfigMapName, "injectConfigMapName", defaultInjectConfigMapName,
+		fmt.Sprintf("ConfigMap name for Istio sidecar injection, key should be %q.", injectConfigMapKey))
+
+	return addToMeshCmd
+}
+
+func svcMeshifyCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "service",
+		Short: "Add Service into Istio Service Mesh",
+		Long: `Find deployments for the service, inject side car and add the service into mesh
```

Comment:
This command doesn't modify any yaml files, it isn't clear why you say it does.

The help should explain what the command is for and what the command does.  Perhaps something like

```
istioctl experimental add-to-mesh restarts pods with the Istio sidecar.  Use 'add-to-mesh'
to test deployments for compatibility with Istio.  If your service does not function after
using 'add-to-mesh' you must re-deploy it and troubleshoot it for Istio compatibility.
See https://istio.io/docs/setup/kubernetes/additional-setup/requirements/
```

Reply:
fixed